Tests/NautilusTelemetryTests/SnapshotAndResetTests.swift

```diff
@@ -0,0 +1,441 @@
+		XCTAssertNil(counter.endTime)
+		XCTAssertGreaterThan(counter.startTime, originalStartTime)
+	}
+
+	// MARK: - Histogram Tests
+
+	func testHistogramSnapshotAndReset() throws {
+		let explicitBounds: [Int] = [10, 50, 100, 500]
+		let histogram = Histogram<Int>(name: "TestHistogram", unit: unit, description: "Test histogram", explicitBounds: explicitBounds)
+
+		// Record some values
+		histogram.record(5, attributes: ["bucket": "small"])
+		histogram.record(75, attributes: ["bucket": "medium"])
+		histogram.record(250, attributes: ["bucket": "large"])
+
+		// Verify histogram has values
+		XCTAssertFalse(histogram.values.values.isEmpty)
+
+		let originalStartTime = histogram.startTime
+
+		// Perform snapshot and reset
+		let snapshot = histogram.snapshotAndReset()
+
+		// Verify snapshot properties
+		XCTAssertEqual(snapshot.name, "TestHistogram")
+		XCTAssertEqual(snapshot.unit?.symbol, "bytes")
+		XCTAssertEqual(snapshot.description, "Test histogram")
+		XCTAssertEqual(snapshot.startTime, originalStartTime)
+		XCTAssertNotNil(snapshot.endTime)
+		XCTAssertEqual(snapshot.aggregationTemporality, .delta)
+
+		// Verify snapshot has the original values
+		guard let snapshotHistogram = snapshot as? Histogram<Int> else {
+			XCTFail("Snapshot should be Histogram<Int>")
+			return
+		}
+		XCTAssertFalse(snapshotHistogram.values.values.isEmpty)
+		XCTAssertEqual(snapshotHistogram.values.explicitBounds, explicitBounds)
+
+		// Verify original histogram is reset
+		XCTAssertTrue(histogram.values.values.isEmpty)
+		XCTAssertNil(histogram.endTime)
+		XCTAssertGreaterThan(histogram.startTime, originalStartTime)
+	}
+
+	// MARK: - ObservableCounter Tests
+
+	func testObservableCounterSnapshotAndReset() throws {
+		var callbackInvoked = false
+		let counter = ObservableCounter<Int>(
+			name: "TestObservableCounter",
+			unit: unit,
+			description: "Test observable counter"
+		) { counter in
+			callbackInvoked = true
+			counter.observe(500, attributes: ["source": "callback"])
+		}
+
+		let originalStartTime = counter.startTime
+
+		// Perform snapshot and reset
+		let snapshot = counter.snapshotAndReset()
+
+		// Verify callback was invoked during snapshot
+		XCTAssertTrue(callbackInvoked)
+
+		// Verify snapshot properties
+		XCTAssertEqual(snapshot.name, "TestObservableCounter")
+		XCTAssertEqual(snapshot.unit?.symbol, "bytes")
+		XCTAssertEqual(snapshot.description, "Test observable counter")
+		XCTAssertEqual(snapshot.startTime, originalStartTime)
+		XCTAssertNotNil(snapshot.endTime)
+		XCTAssertEqual(snapshot.aggregationTemporality, .delta)
+
+		// Verify snapshot has the callback-observed values
+		guard let snapshotCounter = snapshot as? ObservableCounter<Int> else {
+			XCTFail("Snapshot should be ObservableCounter<Int>")
+			return
+		}
+		XCTAssertFalse(snapshotCounter.values.values.isEmpty)
+		XCTAssertTrue(snapshotCounter.values.values.contains { attributes, value in
+			value == 500 && attributes["source"] as? String == "callback"
+		})
```

Comment:
I see that observing sets the value. Can you help me understand how this is intended to be used? I want to learn.

Reply:
I haven't tried using these in practice, but the observable instruments are intended to enable a use case where the data isn't collected until the point of recording.

I think this lets you achieve two things:

- Delay the measurement until the point of reporting. Might be useful for collecting things like memory utilization.
- Avoid the cost of collection if you're not being sampled.